### ISSUE TITLE
Persistent buffers missing in breakpoints when using SecAgg

### ISSUE THREAD

**Issue by sharkovsky** [fedbiomed/fedbiomed#1003]
# Description

Some layers are not being saved correctly in the breakpoint when using SecAgg. 
This leads to a `KeyError` when trying to reload the aggregated parameters from the breakpoint. 

## To reproduce

Modify the `Net`definition in the general-tutorial-secagg notebook to include a batch normalization layer as follows:
```python
    
    class Net(nn.Module):
        def __init__(self, model_args):
            super().__init__()
            self.conv1 = nn.Conv2d(1, 32, 3, 1)
            self.conv2 = nn.Conv2d(32, 64, 3, 1)
            self.dropout1 = nn.Dropout(0.25)
            self.fc1 = nn.Linear(9216, 128)
            self.bn1 = nn.BatchNorm1d(128)
            self.dropout2 = nn.Dropout(0.5)
            self.fc2 = nn.Linear(128, 10)

        def forward(self, x):
            x = self.conv1(x)
            x = F.relu(x)
            x = self.conv2(x)
            x = F.relu(x)
            x = F.max_pool2d(x, 2)
            x = self.dropout1(x)
            x = torch.flatten(x, 1)
            x = self.fc1(x)
            x = F.relu(x)
            x = self.bn1(x)
            x = self.dropout2(x)
            x = self.fc2(x)


            output = F.log_softmax(x, dim=1)
            return output
```

Then save a breakpoint and try to reload the aggregated_params.

**Comment by sharkovsky** [comment]
The problem is in the `TorchTrainingPlan.after_training_params` function. 

This function is quite confusing and almost certainly would cause other bugs. 

The flow of the function is:
1. we first define a `params` variable which is supposed to contain the model parameters. We take into account the user's preference of whether they wanted to shared the persistent buffers or not
2. if a `postprocess` was defined, we recompute `params = self.postprocess(self._model.model.state_dict())`. This is almost certainly wrong, since we probably want to do `params = self.postprocess(params)` instead.
3. we apply Differential Privacy correctly, on the basis of whatever `params` already contained
4. is `flatten` is True, we flatten the parameters. Except that we don't flatten the `params` variable, but we get the parameters directly from the model! 

So in step 2 and in step 4, we potentially discard whatever postprocessing/filtering/modification was done before!

**Comment by sharkovsky** [comment]
# Suggestion for Fix

We should change `after_training_params` such that it takes into account the previous steps each time. 

For step 2, this is quite easy. 
For step 4, I think the best way would be to change `flatten` such that it takes the `params` variable as input.

**Comment by sharkovsky** [comment]
While debugging this, I found that the `share_persistent_buffers` training argument was not correctly propagated. 
I also patched `Job._get_model_params` to be able to propagate this argument (in the form of `only_trainable`, which is actually more generic)

**Comment by sharkovsky** [comment]
This issue is related to #903

**Comment by srcansiz** [comment]
Hi @sharkovsky, 

Do you think this issue is done after merging PR #1035 or it needs some extra work?

**Comment by mvesin** [comment]
hi @lchambon @sharkovsky I don't manage to reproduce the issue running `feature/896-Job-refactor`
- using modified `general-tutorial-secagg.ipynb` as documented
- then `exp.run_once(increase=True)`
- then 
```
loaded_exp = Experiment.load_breakpoint()
loaded_exp.info()
```
- but `loaded_exp.aggregated_params()` does not fail
- plus `(exp.aggregated_params())[0]['params']['bn1.weight']` gives same answer as `(loaded_exp.aggregated_params())[0]['params']['bn1.weight']` -- and same with `bn1.bias`

What am I missing ?

**Comment by sharkovsky** [comment]
Are you using the current branch (specifically, after PR #1035) ?
- this issue was 99% solved by PR #1035 
- while working on #1044  we found some corner cases that were not covered by #1035 (by mistake)
Therefore, we decided to keep both issues open, and close them both when #1044 is merged.

**Comment by mvesin** [comment]
Thank you @sharkovsky @lchambon 

Nice, I could test with `getting-started-tutorial.ipynb` modified as indicated after training & breakpoint we have the same values for
- `loaded_exp.training_plan().get_model_params(exclude_buffers=False)['bn1.running_var']`
- `exp.training_plan().get_model_params(exclude_buffers=False)['bn1.running_var']`

And same for `running_mean` `num_batches_tracked`

```
loaded_exp.training_plan().get_model_params(exclude_buffers=False).keys()
dict_keys(['conv1.weight', 'conv1.bias', 'conv2.weight', 'conv2.bias', 'fc1.weight', 'fc1.bias', 'bn1.weight', 'bn1.bias', 'bn1.running_mean', 'bn1.running_var', 'bn1.num_batches_tracked', 'fc2.weight', 'fc2.bias'])
```
